src/equihash.rs

```diff
@@ -0,0 +1,471 @@
+use blake2_rfc::blake2b::{Blake2b, Blake2bResult};
+use byteorder::{BigEndian, LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::io::Cursor;
+use std::mem::size_of;
+
+struct Params {
+    n: u32,
+    k: u32,
+}
+
+#[derive(Clone)]
+struct Node {
+    hash: Vec<u8>,
+    indices: Vec<u32>,
+}
+
+impl Params {
+    fn indices_per_hash_output(&self) -> u32 {
+        512 / self.n
+    }
+    fn hash_output(&self) -> u8 {
+        (self.indices_per_hash_output() * self.n / 8) as u8
+    }
+    fn collision_bit_length(&self) -> usize {
+        (self.n / (self.k + 1)) as usize
+    }
+    fn collision_byte_length(&self) -> usize {
+        (self.collision_bit_length() + 7) / 8
+    }
+    fn hash_length(&self) -> usize {
+        ((self.k as usize) + 1) * self.collision_byte_length()
+    }
+}
+
+impl Node {
+    fn from_children(a: Node, b: Node, trim: usize) -> Self {
+        let hash: Vec<_> = a.hash
+            .iter()
+            .zip(b.hash.iter())
+            .skip(trim)
+            .map(|(a, b)| a ^ b)
+            .collect();
+        let indices = if a.indices_before(&b) {
+            let mut indices = a.indices;
+            indices.extend(b.indices.iter());
+            indices
+        } else {
+            let mut indices = b.indices;
+            indices.extend(a.indices.iter());
+            indices
+        };
+        Node {
+            hash: hash,
+            indices: indices,
+        }
+    }
+
+    fn from_children_ref(a: &Node, b: &Node, trim: usize) -> Self {
+        let hash: Vec<_> = a.hash
+            .iter()
+            .zip(b.hash.iter())
+            .skip(trim)
+            .map(|(a, b)| a ^ b)
+            .collect();
+        let mut indices = Vec::with_capacity(a.indices.len() + b.indices.len());
+        if a.indices_before(b) {
+            indices.extend(a.indices.iter());
+            indices.extend(b.indices.iter());
+        } else {
+            indices.extend(b.indices.iter());
+            indices.extend(a.indices.iter());
+        }
+        Node {
+            hash: hash,
+            indices: indices,
+        }
+    }
+
+    fn indices_before(&self, other: &Node) -> bool {
+        // Indices are serialized in big-endian so that integer
+        // comparison is equivalent to array comparison
+        self.indices[0] < other.indices[0]
+    }
+
+    fn is_zero(&self, len: usize) -> bool {
+        self.hash.iter().take(len).all(|v| *v == 0)
+    }
+}
+
+fn initialise_state(n: u32, k: u32, digest_len: u8) -> Blake2b {
+    let mut personalization: Vec<u8> = Vec::from("ZcashPoW");
+    personalization.write_u32::<LittleEndian>(n).unwrap();
+    personalization.write_u32::<LittleEndian>(k).unwrap();
+
+    Blake2b::with_params(digest_len as usize, &[], &[], &personalization)
+}
+
+fn generate_hash(base_state: &Blake2b, i: u32) -> Blake2bResult {
+    let mut lei = vec![];
```

Comment:
It would be better to create a fixed-width stack-allocated array here.

```rust
let mut lei = [0u8; 4];
(&mut lei[..]).write_u32::<LittleEndian>(i).unwrap();
```